pycqed/instrument_drivers/meta_instrument/device.py

```diff
@@ -0,0 +1,1032 @@
+import pycqed.measurement.waveform_control.sequence as sequence
+import pycqed.instrument_drivers.meta_instrument.qubit_objects.QuDev_transmon as qdt
+import pycqed.measurement.waveform_control.pulse as bpl
+import scipy.optimize as opti
+from pycqed.analysis import measurement_analysis as ma
+from pycqed.analysis_v3 import pipeline_analysis as pla
+from pycqed.analysis_v3.processing_pipeline import ProcessingPipeline
+from pycqed.measurement import sweep_functions as swf
+from pycqed.measurement.calibration_points import CalibrationPoints
+from pycqed.measurement.sweep_points import SweepPoints
+from pycqed.measurement.waveform_control import pulsar as ps
+from pycqed.utilities import general as gen
+from qcodes.instrument.base import Instrument
+from qcodes.instrument.parameter import (
+    ManualParameter, InstrumentRefParameter)
+from qcodes.utils import validators as vals
+
+log = logging.getLogger(__name__)
+
+
+class Device(Instrument):
+    def __init__(self, name, qubits, connections, **kw):
+        """
+        Instantiates device instrument and adds its parameters.
+
+        Args:
+            name (str): name of the device
+            qubits (list of QudevTransmon): qubits of the device
+            connections: list of elements of the form [qb1, qb2] with qb1 and qb2 QudevTransmon objects. qb1 and qb2
+                         should be physically connected on the device.
+        """
+        super().__init__(name, **kw)
+
+        qb_names = [qb.name for qb in qubits]
+        connectivity_graph = [[qb1.name, qb2.name] for [qb1, qb2] in connections]
+
+        for qb_name in qb_names:
+            setattr(self, qb_name, self.find_instrument(qb_name))
+
+        self._operations = {}  # dictionary containing dictionaries of operations with parameters
+
+        self.add_parameter('qubits',
+                           vals=vals.Lists(),
+                           initial_value=qb_names,
+                           parameter_class=ManualParameter)
+
+        # Instrument reference parameters
+        self.add_parameter('instr_mc',
+                           parameter_class=InstrumentRefParameter)
+        self.add_parameter('instr_pulsar',
+                           parameter_class=InstrumentRefParameter)
+        self.add_parameter('instr_dc_source',
+                           parameter_class=InstrumentRefParameter)
+        self.add_parameter('instr_trigger',
+                           parameter_class=InstrumentRefParameter)
+
+        self.add_parameter('connectivity_graph',
+                           vals=vals.Lists(),
+                           label="Qubit Connectivity Graph",
+                           docstring="Stores the connections between the qubits "
+                                     "in form of a list of lists [qbi, qbj]",
+                           parameter_class=ManualParameter,
+                           initial_value=connectivity_graph
```

Comment:
Agree. Now, allow for both qubit names or object for both the qubits, as well as the connectivity_graph input.

Reply:
I think introducing this connectivity graph is really a good feature :) Though I've more often seen directed acyclic graphs stored in python as vertices as keys, and neighbors a list of neighbors as values,
see e.g. https://www.python.org/doc/essays/graphs/. For 3 qubits in a chain it would be {"qb1": ["qb2"], "qb2": ["qb1", "qb3"], "qb3": ["qb2"]}
Using that representation, we cal also easily find paths through non-neighboring qubits (see attached link), and I think it could become useful when doing swaps based on connectivity.